### ISSUE TITLE
coop sticky algo on large partition numbers

### ISSUE THREAD

**Issue by ericwuseattle** [confluentinc/librdkafka#4629]
Hey there, 

There are 2 issues, I noticed on kafka coop sticky mode.
1. The hard code on partition_cnt inside rd_kafka_sticky_assignor_assign_cb
https://github.com/confluentinc/librdkafka/blob/master/src/rdkafka_sticky_assignor.c#L1834

2. On 3K partitions, it's working without issue, but if I increase the partition to 6K with fresh topic(I mean recreate the topic as new one). Have to increase the session.timeout.ms=10000 and max.poll.interval.ms=10000 from 3s to 10s to make it working.
Otherwise will get kickout out from grp
Broker logs: 
Member XXX-6F958DDF5F-CDXRQ~-0793c679-d5ef-4753-9056-7da314e1415b in group XXX-TOPIC-NAME-XXX has failed, removing it from the group (kafka.coordinator.group.GroupCoordinator). 

I'm not sure what's causing the timeout, but I'm sure we keep calling kafka poll in a timer infinitely. So increase into 10s it's working without any issue. Then I'm trying on 15K partitions with 10s, no lucky, same error.

How to reproduce
On rdkafka 2.3.0
big partition count.
6K partitions with 3s session timeout
or
15K partitions with 10s session timeout

Additionally we were using customize assignment(consistent hash), it's working without any issue with 15K partitions and 10s session timeout.

Thanks in advance,
Eric

**Comment by ericwuseattle** [comment]
Any thoughts on this problem? Are more details needed?

**Comment by emasab** [comment]
@ericwuseattle could you send some logs with `debug=all` . It's possible that it's needed to increase those values for a rebalance with those many partitions, but from the logs we can see where most of the time goes.

**Comment by ericwuseattle** [comment]
unfortunately I do not have the evn was set up to testing in my hand right now, have you checked the hard code of partition count in the code, if we could fix that part first then I'll find time to retry it.

/* FIXME: Let the cgrp pass the actual eligible partition count */
size_t partition_cnt = member_cnt * 10; /* FIXME */

https://github.com/confluentinc/librdkafka/blob/master/src/rdkafka_sticky_assignor.c#L1834

**Comment by emasab** [comment]
@ericwuseattle 
```
/* FIXME: Let the cgrp pass the actual eligible partition count /
size_t partition_cnt = member_cnt * 10; / FIXME */
```
that is just the estimated partition count used for the the initial size of maps and lists. You can try to increase the multiplier and see if it changes something and send some logs of the leader and 2-3 random members.

**Comment by ericwuseattle** [comment]
> Given your configuration you're not using the sticky assignor as `fetch.min.bytes=1, fetch.wait.max.ms=500, fetch.error.backoff.ms=0, heartbeat.interval.ms=1000, enable.auto.commit=false, enable.partition.eof=false, enable.auto.offset.store=false, max.poll.interval.ms=3000, session.timeout.ms=3000`
> 
> as the default `partition.assignment.strategy` doesn't include `cooperative-sticky` could you update your configuration if you're setting it?

We have the setting for
partition.assignment.strategy=cooperative-sticky

**Comment by emasab** [comment]
@ericwuseattle thanks, other helpful info is:
- if you have a rebalance callback set. In that case please test without it
- how many members are there in the group, and if they're all subscribed to that topic with 3-6-15K partitions or to other topics too

**Comment by ericwuseattle** [comment]
> I'm not sure what's causing the timeout, but I'm sure we keep calling kafka poll in a timer infinitely. So increase into 10s it's working without any issue.
> You have to set the callback to do the incremental partitions assing/revoke client api.

> Trying further on 15K partitions with 10s, no lucky, would not work, gets kicked out from grp.
> 30 members totally, only 1 topic.

**Comment by lucas-sonnabend** [comment]
hello,

I think I am running into the same issue here, both at work, and also when I try to reproduce it on my personal laptop.
I can give more detailed information, but the rought set-up is:
- running a 2 broker cluster locally via docker-compose.
- creating a topic with 3000 partitions
- running 450 consumers locally as python threads, using the confluent-kafka library, version 2.8.2, which wraps librdkafka

I can share my whole set-up as a github respository, it is a bit messy, though

this is my consumer configuration:
```

    conf = {
        'bootstrap.servers': bootstrap_servers,
        'group.id': group_id,
        'enable.auto.commit': False,
        'auto.offset.reset': 'earliest',
        'client.id': group_member_id,
        'session.timeout.ms': 15000,
        'heartbeat.interval.ms': 5000,
        'max.poll.interval.ms': 60000,
        'partition.assignment.strategy': 'cooperative-sticky',
        'debug': 'consumer,cgrp,assignor', 
    }
```
The sesstion timeout is set lower than the default to "force the issue".

What I can see in broker logs is the following:

```
21:31:09,315] INFO [GroupCoordinator 2]: Dynamic member with unknown member id joins group test-group in Stable state. Created a new member id extra-0438b5b8-9094-41e5-adc6-2fa2b6f73834 and request the member to rejoin with this id. (kafka.coordinator.group.GroupCoordinator)
21:31:09,324] INFO [GroupCoordinator 2]: Preparing to rebalance group test-group in state PreparingRebalance with old generation 24 (__consumer_offsets-12) (reason: Adding new member extra-0438b5b8-9094-41e5-adc6-2fa2b6f73834 with group instance id None; client reason: not provided) (kafka.coordinator.group.GroupCoordinator)
21:31:11,003] INFO [GroupCoordinator 2]: Stabilized group test-group generation 25 (__consumer_offsets-12) with 451 members (kafka.coordinator.group.GroupCoordinator)

21:31:26,003] INFO [GroupCoordinator 2]: Member thread-consumer-247-1e2c3681-2563-4cd3-ae17-65a212b31b38 in group test-group has failed, removing it from the group (kafka.coordinator.group.GroupCoordinator)
21:31:26,007] INFO [GroupCoordinator 2]: Preparing to rebalance group test-group in state PreparingRebalance with old generation 25 (__consumer_offsets-12) (reason: removing member thread-consumer-247-1e2c3681-2563-4cd3-ae17-65a212b31b38 on heartbeat expiration) (kafka.coordinator.group.GroupCoordinator)
21:31:26,240] INFO [GroupCoordinator 2]: Stabilized group test-group generation 26 (__consumer_offsets-12) with 450 members (kafka.coordinator.group.GroupCoordinator)
21:31:26,400] INFO [GroupCoordinator 2]: Dynamic member with unknown member id joins group test-group in CompletingRebalance state. Created a new member id thread-consumer-247-194b19f5-c420-4a2d-a6dc-cde7ef56c610 and request the member to rejoin with this id. (kafka.coordinator.group.GroupCoordinator)
21:31:26,406] INFO [GroupCoordinator 2]: Preparing to rebalance group test-group in state PreparingRebalance with old generation 26 (__consumer_offsets-12) (reason: Adding new member thread-consumer-247-194b19f5-c420-4a2d-a6dc-cde7ef56c610 with group instance id None; client reason: not provided) (kafka.coordinator.group.GroupCoordinator)
21:31:40,178] INFO [GroupCoordinator 2]: Stabilized group test-group generation 27 (__consumer_offsets-12) with 451 members (kafka.coordinator.group.GroupCoordinator)
21:31:54,062] INFO [GroupCoordinator 2]: Assignment received from leader thread-consumer-211-a74c0469-fb49-4e72-8d1b-a9021b203252 for group test-group for generation 27. The group has 451 members, 0 of which are static. (kafka.coordinator.group.GroupCoordinator)
21:31:54,824] INFO [GroupCoordinator 2]: Preparing to rebalance group test-group in state PreparingRebalance with old generation 27 (__consumer_offsets-12) (reason: Updating metadata for member thread-consumer-367-fbf457ba-86a0-4a15-869c-abee70870c57 during Stable; client reason: not provided) (kafka.coordinator.group.GroupCoordinator)
21:31:59,859] INFO [GroupCoordinator 2]: Stabilized group test-group generation 28 (__consumer_offsets-12) with 451 members (kafka.coordinator.group.GroupCoordinator)
21:32:13,442] INFO [GroupCoordinator 2]: Assignment received from leader thread-consumer-211-a74c0469-fb49-4e72-8d1b-a9021b203252 for group test-group for generation 28. The group has 451 members, 0 of which are static. (kafka.coordinator.group.GroupCoordinator)
```

TLDR: 
-  I create a new consumer to join
-  21:31:11,003 it stabilizes, that means the coordinator asks the leader to compute the new assignment
- 21:31:26,003 (exactly 15s, the session timeout later), the leader "fails" due to a heartbeat timeout
- a new leader is picked, it stabilises, if you are lucky then the leader computes the new assignment fast enough to not get kicked out.

On the leader side I can also get logs (unfortunately from a different run, so the timestamps do not match.

```
%7 2025-03-24 22:13:20.249 STICKY thread-consumer-0#consumer-1  [thrd:main]: Sort 3000 partitions in existing assignment
%7 2025-03-24 22:13:30.689 STICKY thread-consumer-0#consumer-1  [thrd:main]: Prepared balanced reassignment for 451 consumers, 3000 available partition(s) where of 0 are unassigned (initializing=false, revocationRequired=false, 0 fixed assignments)
```
TLDR:
between the sort 3000 partitions and prepared balanced reassignment for 451 consumers, ~10s pass, which is the majority of the time spent re-calculating the assignment.

At work, we have this problem at a slightly higher scale, with 3600 partitions and 1800 consumers, and computing the assignment taking ~4 minutes (I had to increase the session timeout to 6 minutes, which I don't like).

Do these pointers help?

It seems to me that the time spent on the assignment callback grows both with number of partitions, and number of consumers, and it taking minutes feels like an unreasonable amount of time.
Paired with the problem that the leader is not sending heart-beats while running this, it is really annoying

**Comment by lucas-sonnabend** [comment]
for reference, I pushed a repository with my experiments here: https://github.com/lucas-sonnabend/rebalancing-experiments

**Comment by lucas-sonnabend** [comment]
I am currently trying to narrow this down a bit more,
my idea is to get the `rd_kafka_sticky_assignor_assign_cb` running as a unit test with 1000 partitions, and 1000 consumers,
probably based around the `static int ut_testTwoConsumersOneTopicTwoPartitions` test.

I would expect it to take around 60s if my suspicions are correct. (for reference, the coop-sticky algo in the python implementation takes ~1-2 seconds with that many consumers and partition, as a unit test.)

However, I am not that familiar with the unit test structure in C, and how to best run this, so any help here is appreciated.

**Comment by lucas-sonnabend** [comment]
I think for a large count of partitions, this is coming back to haunt me: https://github.com/confluentinc/librdkafka/blob/master/src/rdkafka_partition.c#L3135

**Comment by lucas-sonnabend** [comment]
I'll see if I can get a fix for this in the next couple of days
